shiny/ui/_theme.py

```diff
@@ -0,0 +1,513 @@
+        self._include_paths: list[str] = []
+
+        if isinstance(include_paths, (str, pathlib.Path)):
+            self._include_paths.append(str(include_paths))
+        elif isinstance(include_paths, Sequence):
+            for path in include_paths:
+                self._include_paths.append(str(path))
+
+        # User-provided Sass code
+        self._functions: list[str] = []
+        self._defaults: list[str] = []
+        self._mixins: list[str] = []
+        self._rules: list[str] = []
+
+        # _css is either:
+        # 1. "" indicating that the CSS has not been compiled yet
+        # 2. A string containing the compiled CSS for the current theme
+        self._css: str = ""
+
+        # If the theme has been customized and rendered once, we store the tempdir
+        # so that we can re-use the already compiled CSS file.
+        self._css_temp_srcdir: Optional[str] = None
+
+    @classmethod
+    def available_presets(cls) -> tuple[ShinyThemePreset, ...]:
+        """
+        Get a list of available theme presets.
+        """
+        return shiny_theme_presets
+
+    @property
+    def preset(self) -> ShinyThemePreset:
+        return self._preset
+
+    @preset.setter
+    def preset(self, value: ShinyThemePreset) -> None:
+        check_is_valid_preset(value)
+        self._preset = value
+        self._reset_css()
+
+    def _reset_css(self) -> None:
+        self._css = ""
+        self._css_temp_srcdir = None
+
+    def _has_customizations(self) -> bool:
+        return (
+            len(self._functions) > 0
+            or len(self._defaults) > 0
+            or len(self._mixins) > 0
+            or len(self._rules) > 0
+        )
+
+    def _can_use_precompiled(self) -> bool:
+        return (
+            self._preset in shiny_theme_presets_bundled
+            and not self._has_customizations()
+        )
+
+    def _combine_args_kwargs(
+        self,
+        *args: str,
+        allow_both: bool = True,
+        kwargs: dict[str, str | float | int | bool | None],
+        is_default: bool = False,
+    ) -> list[str]:
+        if not allow_both and len(args) > 0 and len(kwargs) > 0:
+            # Python forces positional arguments to come _before_ kwargs, but default
+            # argument order might matter. To be safe, we force users to pick one order.
+            raise ValueError("Cannot provide both positional and keyword arguments.")
+        if len(args) == 0 and len(kwargs) == 0:
+            return []
+
+        values: list[str] = list(args)
+        default = " !default" if is_default else ""
+
+        if len(kwargs) > 0:
+            for key, value in kwargs.items():
+                key = key.replace("_", "-")
+                if isinstance(value, bool):
+                    value = "true" if value else "false"
+                elif value is None:
+                    value = "null"
+                values.append(f"${key}: {value}{default};")
+
+        return dedent_array(values)
```

Comment:
FWIW, when I read dedent, I don't automatically think of a string, so having the `textwrap` context helps

```suggestion
        return [textwrap.dedent(y) for y in x]
```